Add snapcraft plugin for Qt Build Suite (qbs)

dpniel opened the pull request:
This plugin supports building any applications that use gcc or clang
toolchains, as well as Qt(4&5) based applications.

LP: [#1659537](https://bugs.launchpad.net/snapcraft/+bug/1659537)

3v1n0 commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,146 @@
+
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+        self.build_packages.extend(['qbs', 'gcc', 'clang'])
```
Add Make too

dpniel replied:
Why is Make required here? qbs doesn't use make and handles the execution on the build graph itself.

3v1n0 replied:
Sorry... I misread `/usr/bin/qmake` with `/usr/bin/make` :-)

dpniel replied:
:-) few!

sergiusens commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,146 @@
+      (enum, 'nil' or 'qt4' or 'qt5')
+      Version of Qt to use for qbs profile. If nil the base gcc/clang profile
+      will be used.
+    - profile:
```
does it make sense to call this `qbs-profile`?

I am trying to get to a place where we can somehow namespace things that come from plugins and things that come from core.

sergiusens commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,146 @@
+            'default': 'gcc',
+        }
+
+        schema['properties']['build-variant'] = {
```
this is not documented, also, can this be renamed to `qbs-build-variant`?

sergiusens commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,146 @@
+      will be used.
+    - profile:
+      (enum, 'gcc' or 'clang')
+      Set either gcc or clang as the base build profile
```
mention that gcc is the default please

sergiusens commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,146 @@
+            # a unique'ish name for snap builds. Hopefully this shouldn't clash
+            # with any other local profiles. Each profile should look something
+            # like: profiles.snapcraftqt5clang or profiles.snapcraftnilgcc
+            build_profile = 'snapcraft' + \
```
how about
```python
build_profile = 'snapcraft-qbs-{}-{}'.format(
    self.options.qt_version,
    self.options.profile)
```

and update the comment as well if you take this

sergiusens commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,146 @@
+                            self.options.profile
+
+            # Setup the Qt profile.
+            self.run(['qbs', 'setup-qt', '/usr/bin/qmake', build_profile],
```
since calling qmake on a specific path, shouldn't you before do this:

```
if not os.path.exists('/usr/bin/qmake'):
    raise RuntimeError('The qbs plugin requires qmake on /usr/bin/qmake')
```

Do you also want to hardcode it to that path?

dpniel commented:
@sergiusens Thanks for the review :-) 

I think I have addressed all your comments now.

I also noticed that the integration tests didn't cover the qbs-options property so I have added that in to switch the binary output and make sure it's being passed through correctly. 

Also I'm not sure if having to create a file to fake qmake being on PATH, in order to make the unit test happy is the right thing to do. Is there some better way to fake it's existence on PATH?

sergiusens commented:
@dpniel I'll defer to the QA guy who goes after the call sign @elopio :smile:

elopio commented on integration_tests/snaps/qbs-nil/snapcraft.yaml:
```diff
@@ -0,0 +1,16 @@
+    source: .
+    qbs-options: [
+      'project.printString:true'
+    ]
```
nit comment, so far I haven't seen this style in yaml.
I've seen:

qbs-options: ['project.printString:true']

and 

qbs-options:
  - 'project.printString:true'

Have you seen that usage in other places? If not, please change it.

elopio commented on integration_tests/snaps/qbs-qt/snapcraft.yaml:
```diff
@@ -0,0 +1,16 @@
+    source: .
+    qbs-options: [
+      'project.printString:true'
+    ]
```
ditto: style on the yaml list.

elopio commented on integration_tests/test_qbs_plugin.py:
```diff
@@ -0,0 +1,38 @@
+        binary_output = self.get_output_ignoring_non_zero_exit(
+            os.path.join('stage', 'bin', self.project_dir),
+            cwd=self.project_dir)
+        self.assertEqual("Hello Snapcraft World", binary_output)
```
Please use single quotes, for consistency.

elopio replied:
You are not testing the Wrong snapcraft string message. That would be a nice thing to add here.

elopio commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,157 @@
+            qmake = shutil.which('qmake')
+            if qmake is None:
+                # Let's just check it's not in a common location
+                qmake = os.path.join('usr', 'bin', 'qmake')
```
I don't see a case where qmake is in /usr/bin, but which can't find it.
Do you think this conditional is necessary?

dpniel replied:
Good point. will remove

elopio commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,157 @@
+                qmake = os.path.join('usr', 'bin', 'qmake')
+                if not os.path.exists(qmake):
+                    raise RuntimeError(
+                        'The qbs plugin requires qmake be on PATH')
```
What happens when your machine has both qt5-qmake and qt4-qmake installed? Which one will be /usr/bin/qmake?

dpniel replied:
Hopefully the correct one :-D 

This is handled by setting the QT_SELECT env var which is done in `QbsPlugin._build_environment()`. The qt-chooser magic takes care of the rest at run time, so you should always get the correct qmake.

elopio commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,157 @@
+    def _build_environment(self):
+        env = os.environ.copy()
+        env['QT_SELECT'] = self.options.qt_version
+        return env
```
I like your comments as a guide inside the code blocks. Thanks!

dpniel commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,157 @@
+      (list of strings)
+      Additional options to pass to the qbs invocation.
+    - qt-version:
+      (enum, 'nil' or 'qt4' or 'qt5')
```
For this I was trying to be explicit in the intentions of the plugin. as qbs can be used to build any project and not just Qt apps. I'm happy to remove nil and make it optional if you think it would be clear enough. I'll update the doc string as well to indicate the `qt_version` is optional.

dpniel commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,157 @@
+        schema['properties']['qbs-profile'] = {
+            'enum': ['gcc', 'clang'],
+            'default': 'gcc',
+        }
```
Heh, that's cool! I'm happy to get this right and learn how you guys like it done :-)

dpniel commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,157 @@
+        env = self._build_environment()
+
+        # Setup the toolchains, there will only be gcc or clang by default
+        self.run(['qbs', 'setup-toolchains', '--detect'], env=env)
```
At this point qbs is only interested in finding toolchains it knows about and creating the base profiles for them.

The place where trouble would be caused is in the `qbs config profile.*.baseProfile = 'clang'`. Which would cause the build to fail. But really you would want that to error out at that point, as clang is expected to be in the detected profiles if clang is set in the yaml.

elopio replied:
thanks for the info.

dpniel commented on snapcraft/plugins/qbs.py:
```diff
@@ -0,0 +1,157 @@
+        else:
+            # a unique'ish name for snap builds. Hopefully this shouldn't clash
+            # with any other local profiles. Each profile should look something
+            # like: profiles.snapcraft-qbs-qt5-clang
```
I did think about uuids, but I think it would make debugging profile issues somewhat difficult. You would only really want one profile for each possible configuration in the `~/.config/QtProject/qbs/profiles` directory and a uuid would make it difficult to figure out which profile you need to be looking at. Plus you would end up with multiple duplicates of the same configuration if running snapcraft locally. Which would get messy.

elopio replied:
makes sense.

dpniel commented:
@elopio Regarding you question:

"And a big question: if we find a bug with this qbs plugin, will you be around to give us a hand?"

Yep no problem, just give me a ping. I'm lurking in the background on rocket.u.c/snapcraft or i'm easily found on irc with nick "DanChapman" :-)

elopio commented on integration_tests/test_qbs_plugin.py:
```diff
@@ -0,0 +1,38 @@
+        # The test binaries have the same name as the project dir
+        binary_output = self.get_output_ignoring_non_zero_exit(
+            os.path.join(self.stage_dir, 'bin', self.project_dir))
+        self.assertNotEqual('Wrong snapcraft string', binary_output)
```
sorry, I meant to build them again with project.printString, to make sure that those arguments work.

dpniel replied:
I've changed this so qbs-nil builds with project.printString:false and qbs-qt with project.printString:true.

From what I can see it's not possible to change `qbs-options` from within the test (am i missing something?), and would require adding more test directories with different snapcraft.yaml files. Which seems like it's needlessly duplicating code.

bjorn commented:
@dpniel Would there be any chance of getting this reverted plugin added again? Maybe for some visibility LP: [#1659537](https://bugs.launchpad.net/snapcraft/+bug/1659537) should also be reopened.

bjorn commented:
No idea why it claims I've unassigned @elopio. :man_shrugging:

cjp256 commented:
@bjorn  I would be glad to assist if someone in the community wanted to revive the plugin and iron out any major issues.

bjorn commented:
@cjp256 Alright, I've revived this PR as #2892 with the goal of trying to make those tests pass. Any help would be really appreciated!
